File: web/framework/element.php
<?

/*
This file is part of domserver.

domserver is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

domserver is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with domserver.  If not, see <http://www.gnu.org/licenses/>.
*/

abstract class UIElement
{
    public $tagname = "div";

    function __construct($domserver, $id)
    {
        $this->id = $id;
        $this->appid = "domserver";
        $this->ds = $domserver;
        $this->obj = $this->ds->obj;
        $this->config = $this->ds->config;
        $this->output = $this->ds->output;
        
        $this->output->register_element($this);
        $this->init();
    }
    
    /* Render element HTML tags */
    function render_html($id, $classes, $content)
    {
        if (count($classes))
            $classes = " class=\"" . implode(" ", $classes) . "\"";
        else $classes = "";
        
        return "<div id=\"$id\"$classes>$content</div>\n";
    }
    
    /* Add child element */
    final function add_child($child)
    {
        $this->output->add_op("child", array($this, $child));
    }
    
    /* Remove child element */
    final function remove_child($child)
    {
        $this->output->add_op("unchild", array($this, $child));
    }
    
    /* Swap with sibling element */
    final function swap_with($sibling)
    {
        $this->output->add_op("swap", array($this, $sibling));
    }
    
    /* Set HTML content
        Warning: removes all previously added children.  It should it be called
        _before_ any call to add_child if content _and_ children are wanted.
     */
    final function set_content($html)
    {
        $this->output->add_op("content", array($this, $html));
    }
    
    /* Set DOM property */
    final function set_dom($property, $value)
    {
        $this->output->add_op("dom", array($this, $property, $value));
    }
    
    /* Set CSS property */
    final function set_css($property, $value)
    {
        if (!$this->output)
        {
            $bt = debug_backtrace();
            echo "<pre>";
            var_dump($bt[0]);
            var_dump($bt[1]);
            die("\n\nbt");
        }
        $this->output->add_op("style", array($this, $property, $value));
    }
    
    /* Set CSS class */
    final function set_class($class)
    {
        $this->output->add_op("class", array($this, $class));
    }
    
    /* Unset CSS class */
    final function unset_class($class)
    {
        $this->output->add_op("unclass", array($this, $class));
    }
    
    /* Schedule element update in $interval milliseconds.
        This should be called again in the update() method to reschedule the
        update, when autorefresh is wanted.
    */
    final function schedule_update($interval)
    {
        $this->output->add_op("sched_update", array($this, $interval));
    }
    
    /* Set event handler
        $event is a DOM event (eg. "onclick")
        When the event happens, $target->$method($arg) will be called.
    */
    final function set_handler($event, $target, $method, $arg)
    {
        $this->output->add_op("event", array($this, $event, $target, $method, $arg));
    }
    
    /* Make element draggable
        Dragging the element will show a tooltip containing $label; dropping
        it on a target will send that target the object $objref.
    */
    final function make_draggable($objref, $label)
    {
        $this->output->add_op("drag_src", array($this, $objref, $label));
    }
    
    /* Make element a drag target
        Dropping an object on the element will call $handler_element->$method
        with 2 arguments :
            - this element (the drag target)
            - the object reference ("application:object-id")
    */
    final function make_drag_target($handler_element, $method)
    {
        $this->output->add_op("drag_target", array($this, $handler_element, $method));
    }
    
    private function block_layout($blocks, $columns=TRUE, $overflow="auto")
    {
        $total = 0.0;
        foreach ($blocks as $b) $total += $b[1];
        
        $remaining = 100.0;
        $cnt = count($blocks);
        foreach ($blocks as $b)
        {
            list($element, $weight) = $b;
            $cnt--;
            
            $element->set_css("position", "absolute");
            $element->set_css("overflow", $overflow);
            $element->set_css($columns ? "top" : "left", 0);
            $element->set_css($columns ? "bottom" : "right", 0);
            
            $used = 100.0 - $remaining;
            $element->set_css($columns ? "left" : "top", "$used%");
            
            if ($cnt == 0)
            {
                $remaining = 0;
            }
            else
            {
                $size = 100.0 * $weight / $total;
                $remaining -= $size;
            }
            $element->set_css($columns ? "right" : "bottom", "$remaining%");
        }
    }
    
    /* Generate CSS for a flexible column layout
    
        $columns is an array of column-specs, each of the form:
            array($element, $weight)
            
        Columns will take full height/width of parent element, with relative
        widths computed using column weights (eg. if a column has twice the
        weight of an other, it will be twice as much wide).
    */
    final function column_layout($columns, $overflow="auto")
    {
        $this->block_layout($columns, TRUE, $overflow);
    }
    
    /* Generate CSS for a flexible row layout
        Works exactly like column_layout(), but horizontally.
    */
    final function row_layout($rows, $overflow="auto")
    {
        $this->block_layout($rows, FALSE, $overflow);
    }
    
    final function debug($message)
    {
        $this->output->debug($this, $message);
    }
    
    final function save_data($key, $value)
    {
        $this->ds->save_client_data("{$this->appid}/{$this->id}/$key", $value);
    }
    
    final function load_data($key, $default)
    {
        return $this->ds->load_client_data("{$this->appid}/{$this->id}/$key", $default);
    }
    
    function init() {}
    abstract function render();
    function update() {}
}

class UIImageElement extends UIElement
{
    public $tagname = "img";

    function __construct($domserver, $id, $src)
    {
        parent::__construct($domserver, $id);
        $this->src = $src;
    }
    
    function render_html($id, $classes, $content)
    {
        if (count($classes))
            $classes = " class=\"" . implode(" ", $classes) . "\"";
        else $classes = "";
        
        return "<img id=\"$id\" src=\"{$this->src}\"$classes>\n";
    }
    
    function render() {}
    
    function set_src($src)
    {
        $this->src = $src;
        $this->set_dom("src", $this->src);
    }
}

?>

File: pydomserver/web/framework/element.py
# This file is part of domserver.
#
# domserver is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# domserver is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with domserver.  If not, see <http://www.gnu.org/licenses/>.

from ...errors import ImplementationError

class UIElement:
    
    tagname = "div"
    
    def __init__(self, om, id):
        self.renew(om)
        
        self.id = id
        self.appid = "DOMSERVER"        
        
        self.output.register_element(self)
        self.init()
        
    def renew(self, om):
        self.output = om
        self.rh = om.rh
        self.ui = om.ui
        self.skin = om.ui.skin
        
        self.config = self.rh.context('config')
        self.obj = self.rh.context('obj')
        
    def create(self, cls, id, *args):
        return cls(self.output, id, *args)
        
    def render_html(self, id, classes, content):
        if len(classes):
            classes = ' class="%s"' % ' '.join(classes)
        else:
            classes = "";
            
        return '<%s id="%s"%s>%s</%s>' % (self.tagname, id, classes, content,
            self.tagname)
        
    def add_child(self, child):
        self.output.add_op("child", [self, child])
        child.render()
        
    def remove_child(self, child):
        self.output.add_op("unchild", [self, child])
        
    def swap_with(self, sibling):
        self.output.add_op("swap", [self, sibling])
        
    def set_content(self, html):
        """Set element HTML content.
        
        Warning: removes all previously added children !
        """
        
        if not isinstance(html, (str, unicode)):
            html = str(html)
        self.output.add_op("content", [self, html])
        
    def set_dom(self, prop, value):
        self.output.add_op("dom", [self, prop, value])
        
    def set_css(self, props, pseudoclass=None):
        for p in props:
            self.output.add_op("style", [self, p, props[p], pseudoclass])
        
    def set_class(self, cls):
        self.output.add_op("class", [self, cls])
        
    def unset_class(self, cls):
        self.output.add_op("unclass", [self, cls])
        
    def schedule_update(self, delay_ms):
        self.output.add_op("sched_update", [self, delay_ms])
        
    def set_handler(self, event, handler, arg):
        """Set DOM event handler
        
        'event' is the name of a DOM event (eg. "onclick").  When that event is
        triggered, handler(arg) will be called.
        """
        
        self.output.add_op("event", [self, event, handler, arg])
        
    def set_jshandler(self, event, handler):
        """Set JS function 'handler' as DOM event 'event' handler"""
        
        self.output.add_op("jsevent", [self, event, handler])
        
    def add_jscode(self, code):
        """Execute arbitrary JS code
        
        In 'code', the substrings '{id}' and '{this}' will be replaced by resp.
        the element DOM id and a reference to the element DOM node
        """
        
        self.output.add_op("jscode", [self, code])
        
    def make_draggable(self, objref, label):
        """Make element draggable
        
        Dragging the element will show a tooltip containing label; dropping
        it on a target will send that target the object objref.
        """
        
        self.output.add_op("drag_src", [self, objref, label])
        
    def make_drop_target(self, handler):
        """Make element a drop target
        
        When receiving a draggable element with object 'objref', will call:
            'handler'(self, 'objref')
        """
        
        self.output.add_op("drop_target", [self, handler])
        
    def _block_layout(self, blocks, cols=True, overflow="auto"):
        total = 0.0
        for b in blocks:
            total += b['weight']
            
        rem = 100.0
        cnt = len(blocks)
        for b in blocks:
            element = b['element']
            weight = b['weight']
            cnt -= 1
            
            element.set_css({"position": "absolute"})
            element.set_css({"overflow": overflow})
            element.set_css({"top" if cols else "left": 0})
            element.set_css({"bottom" if cols else "right": 0})
            used = 100.0 - rem
            element.set_css({"left" if cols else "top": "%f%%" % used})
            
            if cnt == 0:
                rem = 0
            else:
                size = 100.0 * weight / total
                rem -= size
                
            element.set_css({"right" if cols else "bottom": "%f%%" % rem})
    
    def column_layout(self, cols, overflow="auto"):
        """Generate CSS for a flexible column layout
    
        'cols' is an array of column specifications which are dicts with keys
            'element' and 'weight'.
            
        Columns will take full height/width of parent element, with relative
        widths computed using column weights (eg. if a column has twice the
        weight of an other, it will be twice as wide).
        """
        
        self._block_layout(cols, True, overflow)
        
    def row_layout(self, rows, overflow="auto"):
        """Generate CSS for a flexible row layout
        
        Works exactly like column_layout(), but horizontally.
        """
        
        self._block_layout(rows, False, overflow)
        
    def debug(self, message):
        self.output.debug(self, message)
        
    def save(self, key, value):
        self.rh.set_client_data("%s/%s/%s" % (self.appid, self.id, key), value)
        
    def load(self, key, default):
        return self.rh.get_client_data(
            "%s/%s/%s" % (self.appid, self.id, key),
            default
        )
            
    def init(self):
        pass
        
    def update(self):
        pass
        
    def render(self):
        raise ImplementationError("UIElement.render() not overriden")
        
        
class UIImageElement(UIElement):
    tagname = "img"
    
    def __init__(self, om, id, src):
        UIElement.__init__(self, om, id)
        self.src = src
        
    def render_html(self, id, classes, cnt):
        if len(classes):
            classes = ' class="%s"' % ' '.join(classes)
        else:
            classes = ''
            
        return '<img id="%s" src="%s"%s>' % (id, self.src, classes)
        
    def render(self):
        self.set_dom("src", self.src)
        
    def set_src(self, src):
        self.src = src
        self.set_dom("src", self.src)
        
